src/common/hooks/useDownloadTrackButtons.ts

```diff
@@ -1,44 +1,80 @@
-import { useCallback } from 'react'
-
 import moment from 'moment'
-import { useSelector, shallowEqual, useDispatch } from 'react-redux'
-import { useLocation } from 'react-router-dom'
+import { useSelector as reduxUseSelector, shallowEqual } from 'react-redux'
 
 import { ID } from 'common/models/Identifiers'
 import { stemCategoryFriendlyNames, StemCategory } from 'common/models/Stems'
 import { Track, StemTrack } from 'common/models/Track'
+import { CommonState } from 'common/store'
 import { getHasAccount } from 'common/store/account/selectors'
 import { getTrack, getTracks } from 'common/store/cache/tracks/selectors'
-import {
-  openSignOn,
-  updateRouteOnExit,
-  updateRouteOnCompletion,
-  showRequiresAccountModal
-} from 'pages/sign-on/store/actions'
-import { getCurrentUploads } from 'store/application/ui/stemsUpload/selectors'
-import { AppState } from 'store/types'
-
-import {
-  ButtonState,
-  messages,
-  DownloadButtonProps,
-  ButtonType
-} from './DownloadButtons'
+import { getCurrentUploads } from 'common/store/stems-upload/selectors'
+
+export type DownloadButtonConfig = {
+  state: ButtonState
+  type: ButtonType
+  label: string
+  onClick?: () => void
+}
+
+export enum ButtonState {
+  PROCESSING,
+  LOG_IN_REQUIRED,
+  DOWNLOADABLE,
+  REQUIRES_FOLLOW
+}
+
+export enum ButtonType {
+  STEM,
+  TRACK
+}
+
+type Stem = {
+  category: StemCategory
+  downloadable: boolean
+  downloadURL?: string
+  id?: ID
+}
+
+type LabeledStem = Omit<Stem, 'category'> & { label: string }
+
+type UseDownloadTrackButtonsArgs = {
+  following: boolean
+  isOwner: boolean
+  onDownload: (
+    trackID: number,
+    cid: string,
+    category?: string,
+    parentTrackId?: ID
+  ) => void
+  onNotLoggedInClick?: () => void
+}
+
+const messages = {
+  getDownloadTrack: (stemCount: number) => `${stemCount ? 'Original' : ''}`,
+  getDownloadStem: (friendlyName: string, categoryCount: number) =>
+    `${friendlyName} ${categoryCount || ''}`
+}
 
 const doesRequireFollow = (
   isOwner: boolean,
   following: boolean,
   track: Track
 ) => !isOwner && !following && track.download?.requires_follow
 
-const useCurrentStems = (trackId: ID) => {
+const useCurrentStems = ({
+  trackId,
+  useSelector
+}: {
+  trackId: ID
+  useSelector: typeof reduxUseSelector
```

Comment:
not sure if this makes sense, but could we do a runtime/compile-time check for which env we are, and use `useSelector` or `useSelectorWeb`? would probably mean moving `useSelectorWeb` over from `audius-mobile-client` or at least creating a duplicate?

Reply:
Passing `useSelector` in here isn't really for the purpose of using useSelectorWeb in mobile. I had to do this because of the weird issue where calling an external hook that calls a redux hook causes a `"could not find react-redux context value"` error. Passing in `useSelector` and/or `dispatch` alleviates this. I think the better solution would be to get all of our dependencies in sync like outlined here: https://github.com/reduxjs/react-redux/issues/1489


Apart from that, your question is still valid though. I think there could be cases where we want to use the same hook logic from mobile but sometimes with useSelector and sometimes with useSelector. So probably best not to bake that logic into the hook itself. Also that logic would go away when we finally sever the dependency on the web client

Reply:
ohhh i see i see, yeah would be down to look into the version issues at some point, just lmk. and yeah prob just hold off then until we remove dependency on web client